src/Core/Managed/Shared/Extensibility/Implementation/Metrics/DependencyMetricsExtractor.cs

```diff
@@ -0,0 +1,361 @@
+                            //// In case of very high contention, this may happen a little more often,
+                            //// but will no longer happen once the MaxDependencyTypesToDiscover limit is reached.
+                            try
+                            {
+                                typeMetrics = thisMetrics.ByType.GetOrAdd(
+                                        dependencyType,
+                                        (depType) =>
+                                        {
+                                            lock (thisMetrics.TypeDiscoveryLock)
+                                            {
+                                                if (thisMetrics.DependencyTypesDiscoveredCount >= thisMetrics.MaxDependencyTypesToDiscover)
+                                                {
+                                                    throw new InvalidOperationException("MaxDependencyTypesToDiscover reached.");
+                                                }
+
+                                                thisMetrics.DependencyTypesDiscoveredCount++;
+
+                                                return new SucceessAndFailureMetrics(
+                                                    thisMetricManager.CreateMetric(
+                                                            MetricTerms.Autocollection.MetricNames.DependencyCall.Duration,
+                                                            new Dictionary<string, string>()
+                                                            {
+                                                                [MetricTerms.Autocollection.DependencyCall.PropertyNames.Success] = Boolean.TrueString,  // SUCCESS metric
+                                                                [MetricTerms.Autocollection.DependencyCall.PropertyNames.TypeName] = depType,
+                                                            }),
+                                                    thisMetricManager.CreateMetric(
+                                                            MetricTerms.Autocollection.MetricNames.DependencyCall.Duration,
+                                                            new Dictionary<string, string>()
+                                                            {
+                                                                [MetricTerms.Autocollection.DependencyCall.PropertyNames.Success] = Boolean.FalseString, // FAILURE metric
+                                                                [MetricTerms.Autocollection.DependencyCall.PropertyNames.TypeName] = depType,
+                                                            }));
+                                            }
+                                        });
+                            }
+                            catch (InvalidOperationException)
```

Comment:
It looks like we are using exceptions to control execution flow, is there some other pattern we can use here to achieve the same result without the unnecessary exception?